File: bakerysense-web/src/lib/hierarchical.ts
/**
 * Hierarchical reconciliation for multi-level forecasts.
 *
 * Problem: SKU-level forecasts produced independently don't sum to the
 * category-level forecast or the tenant-total forecast. Operators see
 * incoherent numbers ("we predict 130 baguettes for branch A but the
 * BREAD-category total is 125 — which is right?").
 *
 * Solution: project base forecasts onto the coherent subspace of the
 * hierarchy. Two methods supported:
 *
 * 1. **Bottom-up** — trust only the leaves; aggregate upward. Simple,
 *    optimal when leaf forecasts are unbiased and uncorrelated. Loses
 *    information that higher-level forecasts may carry.
 *
 * 2. **OLS-MinT** — solve a least-squares projection that minimises the
 *    sum of squared adjustments to all levels jointly. Produces coherent
 *    forecasts that respect both the leaf models AND any direct higher-
 *    level signal (e.g. seasonal totals from a separate FM).
 *
 * V1.5 uses bottom-up by default — we don't yet have separate higher-
 * level forecasters. When V2 lands, OLS-MinT becomes valuable: TimesFM
 * can produce a category-level forecast directly (its context window is
 * shorter than per-SKU sales but the aggregate is denser), and we
 * reconcile against the SKU-level prediction.
 *
 * Reference: Hyndman et al., "Optimal forecast reconciliation for
 * hierarchical and grouped time series", Computational Statistics & Data
 * Analysis 97 (2016).
 */

export interface Node {
  id: string;
  /** Children IDs. Empty array = leaf. */
  children: string[];
}

export interface Hierarchy {
  /** Root node ID (the tenant total). */
  root: string;
  /** All nodes keyed by id. Must include root and all leaves and every
   *  intermediate level. */
  nodes: Record<string, Node>;
}

/** Walks the tree from root and returns nodes in post-order (children
 *  before parents) — useful for bottom-up aggregation. */
export function postOrder(h: Hierarchy): string[] {
  const out: string[] = [];
  const visit = (id: string): void => {
    const n = h.nodes[id];
    if (!n) return;
    for (const c of n.children) visit(c);
    out.push(id);
  };
  visit(h.root);
  return out;
}

export function leaves(h: Hierarchy): string[] {
  return Object.values(h.nodes)
    .filter((n) => n.children.length === 0)
    .map((n) => n.id);
}

/**
 * Bottom-up reconciliation: take leaf forecasts, sum upward into all
 * ancestors. Any base forecast supplied for non-leaves is overwritten.
 */
export function bottomUp(
  h: Hierarchy,
  base: Record<string, number>,
): Record<string, number> {
  const out: Record<string, number> = {};
  for (const id of postOrder(h)) {
    const n = h.nodes[id];
    if (n.children.length === 0) {
      out[id] = base[id] ?? 0;
    } else {
      let sum = 0;
      for (const c of n.children) sum += out[c] ?? 0;
      out[id] = sum;
    }
  }
  return out;
}

/**
 * OLS-MinT reconciliation. Solves the closed form:
 *
 *   reconciled = S (S' S)^-1 S' base
 *
 * where S is the summing matrix mapping leaves → all levels and base is
 * the column vector of base forecasts at every node.
 *
 * Implementation note: for typical bakery hierarchies (≤ a few hundred
 * leaves) this is fine to compute in pure JS without a BLAS dependency.
 * If we ever forecast against thousands of leaves, port to wasm-blas.
 */
export function olsMinT(
  h: Hierarchy,
  base: Record<string, number>,
): Record<string, number> {
  const allIds = postOrder(h);
  const leafIds = leaves(h);
  const ancestorsOf = (leaf: string): string[] => {
    const out: string[] = [];
    for (const id of allIds) {
      const n = h.nodes[id];
      if (containsLeaf(h, n, leaf)) out.push(id);
    }
    return out;
  };

  // Build the FULL summing matrix S_full (m × n): rows = all nodes,
  // cols = leaves. S_full[i,j] = 1 iff node i is an ancestor of (or
  // equals) leaf j.
  const m = allIds.length;
  const n = leafIds.length;
  const S_full: number[][] = Array.from({ length: m }, () => Array(n).fill(0));
  const leafIndex = new Map(leafIds.map((id, j) => [id, j]));
  const nodeIndex = new Map(allIds.map((id, i) => [id, i]));
  for (const leaf of leafIds) {
    const j = leafIndex.get(leaf)!;
    for (const anc of ancestorsOf(leaf)) {
      const i = nodeIndex.get(anc);
      if (i != null) S_full[i][j] = 1;
    }
  }

  // OLS-MinT projects observed base forecasts onto the coherent
  // subspace. "Observed" means a base forecast was actually supplied —
  // missing values must NOT be treated as zero (that would pull every
  // unobserved level toward zero). Filter S to the rows where base is
  // defined; if only leaves are observed, the projection collapses to
  // the identity (== bottom-up).
  const observed: { id: string; rowIdx: number; value: number }[] = [];
  for (const id of allIds) {
    if (base[id] !== undefined) {
      observed.push({ id, rowIdx: nodeIndex.get(id)!, value: base[id] });
    }
  }

  // Defensive fallback — caller passed nothing useful; emit zeros.
  if (observed.length === 0) {
    const out: Record<string, number> = {};
    for (const id of allIds) out[id] = 0;
    return out;
  }

  const S_obs = observed.map((o) => S_full[o.rowIdx]);
  const y_obs = observed.map((o) => o.value);

  // Compute G = (S_obs' S_obs)^-1 S_obs'   →  shape (n × |observed|)
  const StS = matMul(transpose(S_obs), S_obs);   // n × n
  const StSinv = invertSmallMatrix(StS);          // n × n
  const G = matMul(StSinv, transpose(S_obs));     // n × |observed|

  // Reconciled bottom-level forecasts: b = G y_obs
  const b = matVec(G, y_obs);

  // Lift back to all levels via the FULL summing matrix.
  const r = matVec(S_full, b);

  const out: Record<string, number> = {};
  for (let i = 0; i < m; i++) out[allIds[i]] = r[i];
  return out;
}

// ──── small linear-algebra helpers (no external deps) ────────────────────

function transpose(A: number[][]): number[][] {
  const r = A.length;
  const c = A[0]?.length ?? 0;
  const out: number[][] = Array.from({ length: c }, () => Array(r).fill(0));
  for (let i = 0; i < r; i++) for (let j = 0; j < c; j++) out[j][i] = A[i][j];
  return out;
}

function matMul(A: number[][], B: number[][]): number[][] {
  const r = A.length;
  const k = A[0]?.length ?? 0;
  const c = B[0]?.length ?? 0;
  const out: number[][] = Array.from({ length: r }, () => Array(c).fill(0));
  for (let i = 0; i < r; i++) {
    for (let j = 0; j < c; j++) {
      let s = 0;
      for (let l = 0; l < k; l++) s += A[i][l] * B[l][j];
      out[i][j] = s;
    }
  }
  return out;
}

function matVec(A: number[][], x: number[]): number[] {
  const r = A.length;
  const c = A[0]?.length ?? 0;
  const out = Array(r).fill(0);
  for (let i = 0; i < r; i++) {
    let s = 0;
    for (let j = 0; j < c; j++) s += A[i][j] * x[j];
    out[i] = s;
  }
  return out;
}

/** Gauss-Jordan inverse for small symmetric positive-definite matrices.
 *  Caller is expected to know the matrix is invertible (it is for any
 *  non-degenerate hierarchy). */
function invertSmallMatrix(A: number[][]): number[][] {
  const n = A.length;
  const M: number[][] = A.map((row, i) => [
    ...row,
    ...Array(n).fill(0).map((_, j) => (i === j ? 1 : 0)),
  ]);
  for (let i = 0; i < n; i++) {
    // Pivot selection
    let pivot = i;
    for (let k = i + 1; k < n; k++) if (Math.abs(M[k][i]) > Math.abs(M[pivot][i])) pivot = k;
    if (pivot !== i) [M[i], M[pivot]] = [M[pivot], M[i]];
    const div = M[i][i];
    if (Math.abs(div) < 1e-12) throw new Error("singular matrix in olsMinT");
    for (let j = 0; j < 2 * n; j++) M[i][j] /= div;
    for (let k = 0; k < n; k++) {
      if (k === i) continue;
      const factor = M[k][i];
      for (let j = 0; j < 2 * n; j++) M[k][j] -= factor * M[i][j];
    }
  }
  return M.map((row) => row.slice(n));
}

function containsLeaf(h: Hierarchy, n: Node, leaf: string): boolean {
  if (n.id === leaf) return true;
  for (const c of n.children) {
    const child = h.nodes[c];
    if (!child) continue;
    if (containsLeaf(h, child, leaf)) return true;
  }
  return false;
}

// ──── builders for typical bakery hierarchies ────────────────────────────

/**
 * Build a tenant → branch → SKU hierarchy. Useful when you have base
 * forecasts at every level and want to reconcile across branches and
 * SKUs simultaneously.
 */
export function buildTenantHierarchy(
  tenantId: string,
  branchSkus: Record<string, string[]>,  // branchId → SKU IDs at that branch
): Hierarchy {
  const nodes: Record<string, Node> = {
    [tenantId]: { id: tenantId, children: Object.keys(branchSkus) },
  };
  for (const [branchId, skus] of Object.entries(branchSkus)) {
    nodes[branchId] = { id: branchId, children: skus };
    for (const sku of skus) {
      nodes[sku] = { id: sku, children: [] };
    }
  }
  return { root: tenantId, nodes };
}

File: src/bakerysense/hierarchical.py
"""Hierarchical reconciliation for multi-level forecasts.

Python port of bakerysense-web/src/lib/hierarchical.ts (Sprint 4).
Same semantics, numpy for the linear algebra.

Two methods:

  1. bottom_up — trust leaves, aggregate upward. Optimal when leaf
     forecasts are unbiased and uncorrelated.

  2. ols_mint — closed-form OLS projection onto the coherent subspace.
     Reconciles base forecasts at every observed level into a coherent
     set. Filters S to observed rows so missing forecasts at higher
     levels aren't treated as zero (a real bug fixed in the TS port).

Reference: Hyndman et al., "Optimal forecast reconciliation for
hierarchical and grouped time series", CSDA 97 (2016).
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np


@dataclass(frozen=True)
class Hierarchy:
    """Tree as (root_id, {node_id: [child_ids]})."""
    root: str
    nodes: dict[str, list[str]]

    def post_order(self) -> list[str]:
        """Children before parents — for bottom-up sums."""
        out: list[str] = []
        def visit(node_id: str) -> None:
            for c in self.nodes.get(node_id, []):
                visit(c)
            out.append(node_id)
        visit(self.root)
        return out

    def leaves(self) -> list[str]:
        return [nid for nid, kids in self.nodes.items() if not kids]

    def descendants(self, node_id: str) -> set[str]:
        """All leaves under a given node."""
        if node_id not in self.nodes:
            return set()
        kids = self.nodes[node_id]
        if not kids:
            return {node_id}
        out: set[str] = set()
        for c in kids:
            out |= self.descendants(c)
        return out


def bottom_up(h: Hierarchy, base: dict[str, float]) -> dict[str, float]:
    """Take leaf forecasts in `base`; compute every ancestor as the sum
    of its children. Any base value at a non-leaf is overwritten."""
    out: dict[str, float] = {}
    for nid in h.post_order():
        kids = h.nodes.get(nid, [])
        if not kids:
            out[nid] = float(base.get(nid, 0.0))
        else:
            out[nid] = float(sum(out.get(c, 0.0) for c in kids))
    return out


def ols_mint(h: Hierarchy, base: dict[str, float]) -> dict[str, float]:
    """OLS-MinT closed-form reconciliation:

        b̂  = (Sᵀ_obs S_obs)⁻¹ Sᵀ_obs y_obs
        ŷ  = S_full b̂

    where S_full maps leaves → all nodes (rows = nodes, cols = leaves;
    1 if node is an ancestor of leaf, else 0). S_obs is S_full filtered
    to rows for which `base` actually has a forecast — missing higher-
    level forecasts are NOT zeroed (which would bias every unobserved
    level toward zero)."""
    all_ids = h.post_order()
    leaf_ids = h.leaves()
    n_nodes = len(all_ids)
    n_leaves = len(leaf_ids)

    leaf_index = {lid: j for j, lid in enumerate(leaf_ids)}
    node_index = {nid: i for i, nid in enumerate(all_ids)}

    # S_full: (n_nodes, n_leaves), S[i,j]=1 iff node i is an ancestor of leaf j
    S_full = np.zeros((n_nodes, n_leaves))
    for nid in all_ids:
        i = node_index[nid]
        for desc in h.descendants(nid):
            j = leaf_index.get(desc)
            if j is not None:
                S_full[i, j] = 1.0

    # Filter to observed rows
    observed_idx = [node_index[nid] for nid in all_ids if nid in base]
    if not observed_idx:
        return {nid: 0.0 for nid in all_ids}

    S_obs = S_full[observed_idx, :]
    y_obs = np.array([float(base[all_ids[i]]) for i in observed_idx])

    # b̂ = (S_obsᵀ S_obs)⁻¹ S_obsᵀ y_obs  via least-squares for stability
    b_hat, *_ = np.linalg.lstsq(S_obs, y_obs, rcond=None)
    r = S_full @ b_hat
    return {all_ids[i]: float(r[i]) for i in range(n_nodes)}


def flat_hierarchy(leaves: list[str], root: str = "TOTAL") -> Hierarchy:
    """Single-level (root → leaves) hierarchy. Useful when you only have
    a tenant total and per-SKU leaves with no intermediate categories."""
    nodes = {root: list(leaves)}
    for lid in leaves:
        nodes[lid] = []
    return Hierarchy(root=root, nodes=nodes)


def two_level_hierarchy(
    families: dict[str, list[str]],
    root: str = "TOTAL",
) -> Hierarchy:
    """root → family → leaves. `families` maps family_name → [sku_ids]."""
    nodes: dict[str, list[str]] = {root: list(families.keys())}
    for fam, skus in families.items():
        nodes[fam] = list(skus)
        for sku in skus:
            nodes[sku] = []
    return Hierarchy(root=root, nodes=nodes)
